### PR TITLE
Refactor ote arg converters

### PR DESCRIPTION
**Pull request by LeonidBeynenson**

### CODE REVIEW

**Comment by LeonidBeynenson** [comment]
Please, note that I squashed all my changes to one commit.
If you want to review this change as a sequence of commits, see my auxiliary branch
https://github.com/openvinotoolkit/training_extensions/tree/lb/refactor_arg_converters_2__ver0

(I will remove the auxiliary branch from the repo when this PR is merged)

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 1]
```diff
@@ -15,97 +15,184 @@
...
+        pass
 
-        template_folder = os.path.dirname(model_template_path)
-        converted_args = {
-            'config': os.path.join(template_folder, args['config']),
-            'out': os.path.join(args['save_checkpoints_to'], model_template_path),
-            'update_config': update_args,
-        }
-        converted_args.update(self.__map_args(args, self.train_out_args_map))
+    @abstractmethod
```
Comment (evgeny-izutov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:63]: Is it really necessary to declare `abstractmethod` here if most of `arg_converters` will use "as is"? We can reduce copy-paste by removing it.

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 2]
```diff
@@ -15,97 +15,184 @@
...
+            from: the name of cmd line arguments of the corresponding **train** argparser.parse_args
+            to:   the names of additional parameters of the `__call__` method of the corresponding ote trainer class
+                  (e.g. MMDetectionTrainer) in the form suitable for `trainer_class(**kwargs)`
+        """
+        return {
+                'gpu_num': 'gpu_num',
+                'tensorboard_dir': 'tensorboard_dir'
+               }
+    @abstractmethod
+    def compress_out_args_map(self):
```
Comment (evgeny-izutov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:76]: ditto

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 3]
```diff
@@ -15,97 +15,184 @@
...
+        return {
+                'gpu_num': 'gpu_num',
+                'tensorboard_dir': 'tensorboard_dir',
+                'nncf_quantization': 'nncf_quantization',
+                'nncf_pruning': 'nncf_pruning',
+                'nncf_sparsity': 'nncf_sparsity',
+                'nncf_binarization': 'nncf_binarization',
+               }
+    @abstractmethod
+    def test_out_args_map(self):
```
Comment (evgeny-izutov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:95]: ditto

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 4]
```diff
@@ -15,97 +15,184 @@
...
+                  evaluator class (e.g. MMDetectionEvaluator) in the form suitable for
+                  call `evaluator_class(**kwargs)`
+        """
+        return {
+                'load_weights': 'snapshot',
+                'save_metrics_to': 'out',
+                'save_output_to': 'show_dir'
+               }
+
+    @abstractmethod
```
Comment (evgeny-izutov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:109]: ditto

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 5]
```diff
@@ -15,97 +15,184 @@
...
+    def get_extra_train_args(self, args):
+        """ Gets from the parsed output of the corresponding ote train argparser.parse_args
+            (e.g. `argparser = DefaultArgParser.get_train_parser()`)
+            additional changes that should be applied to mmdetection/mmaction/other config file,
+            the changes will be in the form compatible with mmcv.Config.merge_from_dict
+        """
+        return {}
 
-        return converted_args
+    @abstractmethod
```
Comment (evgeny-izutov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:118]: ditto

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 6]
```diff
@@ -15,97 +15,184 @@
...
+        """ Gets from the parsed output of the corresponding ote test argparser.parse_args
+            (e.g. `argparser = DefaultArgParser.get_test_parser()`)
+            additional changes that should be applied to mmdetection/mmaction/other config file,
+            the changes will be in the form compatible with mmcv.Config.merge_from_dict
+        """
+        return {}
 
-    def convert_compress_args(self, model_template_path, args):
-        update_args = self.__map_args(args, self.compress_update_args_map)
+    @abstractmethod
```
Comment (evgeny-izutov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:127]: ditto

File: pytorch_toolkit/ote/ote/modules/arg_converters/mmdetection.py [hunk 7]
```diff
@@ -128,7 +144,7 @@ def _get_extra_train_args(self, args):
 
         return classes_list_to_update_config_dict(args['config'], classes)
```
Comment (evgeny-izutov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/mmdetection.py:145]: `return self.get_extra_test_args(args)` or move it into separate function to exclude copy-paste?

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 1 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py:63]: I like to be sure that nobody forget to implement/override something.
So, I force the person who makes a derived class to implement all of the required methods.

IMHO, the copy of code is small -- just one line per function.
But if Ilya also supposes that it is not required -- OK, I'll remove `@abstractmethod` here

File: pytorch_toolkit/ote/ote/modules/arg_converters/mmdetection.py (hunk 7 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/mmdetection.py:145]: Evgeny, I copied the functions from the previous implementation of `MMDetectionCustomClassesArgConverter` made by Ilya earlier and I did not change them. I would prefer to move refactoring of the functions out of scope of this PR.

Please, note that for train and test the functions do a bit different things (different work with `classes_from_snapshot`, etc).

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 1 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py:63]: 2 @Ilya-Krylov : Ilya, what is your opinion?

**Comment by LeonidBeynenson** [comment]
Merged branch 'develop' into `lb/refactor_arg_converters_2`

2 @sovrasov : Vlad, I merged your changes into my refactoring of arg converters.

File: pytorch_toolkit/ote/ote/modules/arg_converters/mmdetection.py (hunk 7 above)
Comment (evgeny-izutov) [reply pytorch_toolkit/ote/ote/modules/arg_converters/mmdetection.py:145]: Sorry, but my comment is mostly about a couple lines above, which are the same for `get_extra_train_args`, `get_extra_test_args`, `get_extra_compress_args` methods.

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 1 above)
Comment (evgeny-izutov) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py:63]: Anyway if somebody know what is he doing he also know (can find) where to change it. The rest people don't need to re-implement each line of base class...

Comment (Ilya-Krylov) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py:63]: @LeonidBeynenson I agree with @evgeny-izutov

Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py:63]: Ok!

Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py]: Fixed

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 2 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py:75]: Fixed

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 3 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py:94]: Fixed

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 4 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py]: Fixed

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 5 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py]: Fixed

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 6 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py]: Fixed

File: pytorch_toolkit/ote/ote/modules/arg_converters/mmdetection.py (hunk 7 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/mmdetection.py:126]: No, they are a bit different -- different text of the messages, `logging.warning` vs `raise RuntimeError`, `get_extra_test_args` contains loading the `classes_from_snapshot` from `'.extra_params.yml'` file (whereas for train case it MUST NOT be done), etc

I agree that some common part may be extracted, but 
* different part of message texts will have to be passed to the common part
* I would like to move refactoring of the code outside the scope of this PR

Comment (evgeny-izutov) [reply pytorch_toolkit/ote/ote/modules/arg_converters/mmdetection.py:126]: Ok, but not forget to add refactoring in the future to make the code great again.

**Review by evgeny-izutov — APPROVED** [review]

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 8]
```diff
@@ -15,97 +15,183 @@
...
+
+class ArgConverterMaps(metaclass=ABCMeta):
+    @abstractmethod
+    def train_update_args_map(self):
+        """
+        Returns a map:
+            from: the name of cmd line arguments of the corresponding **train** argparser.parse_args
+                  (e.g. `argparser = DefaultArgParser.get_train_parser()`)
+            to:   the fields of mmdetection/mmaction/other config for its **train** script in the form
+                  compatible with mmcv.Config.merge_from_dict
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:33]: mmcv.Config is not always config for any algo

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 9]
```diff
@@ -15,97 +15,183 @@
...
-    def convert_train_args(self, model_template_path, args):
-        update_args = self.__map_args(args, self.train_update_args_map)
+    @abstractmethod
+    def test_update_args_map(self):
+        """
+        Returns a map:
+            from: the name of cmd line arguments of the corresponding **test** argparser.parse_args
+                  (e.g. `argparser = DefaultArgParser.get_test_parser()`)
+            to:   the fields of mmdetection/mmaction/other config for its **test** script in the form
+                  compatible with mmcv.Config.merge_from_dict
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:44]: ditto

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 10]
```diff
@@ -15,97 +15,183 @@
...
-        extra_args = self._get_extra_train_args(args)
-        update_args.update(extra_args)
+    @abstractmethod
+    def compress_update_args_map(self):
+        """
+        Returns a map:
+            from: the name of cmd line arguments of the corresponding **compress** argparser.parse_args
+                  (e.g. `argparser = DefaultArgParser.get_compression_parser()`)
+            to:   the fields of mmdetection/mmaction/other config for its **train** script in the form
+                  compatible with mmcv.Config.merge_from_dict to makes compression
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:55]: ditto

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 11]
```diff
@@ -15,97 +15,183 @@
...
 
-        return converted_args
 
-    def convert_compress_args(self, model_template_path, args):
-        update_args = self.__map_args(args, self.compress_update_args_map)
+    def get_extra_test_args(self, args):
+        """ Gets from the parsed output of the corresponding ote test argparser.parse_args
+            (e.g. `argparser = DefaultArgParser.get_test_parser()`)
+            additional changes that should be applied to mmdetection/mmaction/other config file,
+            the changes will be in the form compatible with mmcv.Config.merge_from_dict
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:122]: ditto

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py [hunk 12]
```diff
@@ -15,97 +15,183 @@
...
-            'config': os.path.join(template_folder, args['config']),
-            'out': os.path.join(args['save_checkpoints_to'], model_template_path),
-            'update_config': update_args,
-        }
-        converted_args.update(self.__map_args(args, self.compress_out_args_map))
+    def get_extra_compress_args(self, args):
+        """ Gets from the parsed output of the corresponding ote compress argparser.parse_args
+            (e.g. `argparser = DefaultArgParser.get_compression_parser()`)
+            additional changes that should be applied to mmdetection/mmaction/other config file,
+            the changes will be in the form compatible with mmcv.Config.merge_from_dict
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/ote/ote/modules/arg_converters/base.py:131]: ditto

**Review by Ilya-Krylov — CHANGES_REQUESTED** [review]

**Comment by Ilya-Krylov** [comment]
@LeonidBeynenson As far as I can see derivatives of BaseArgConverter does nothing in all cases. Do we need these derived classes? What about registering derivatives of ArgConverterMaps instead of derivatives of BaseArgConverter?

**Comment by LeonidBeynenson** [comment]
2 @Ilya-Krylov , @evgeny-izutov :  Made changes w.r.t. our considerations.
Please, check if everything is OK now.

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 8 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py:33]: Wrote explanation

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 9 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py:51]: Wrote explanation

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 10 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py]: Wrote explanation

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 11 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py]: Wrote explanation

File: pytorch_toolkit/ote/ote/modules/arg_converters/base.py (hunk 12 above)
Comment (LeonidBeynenson) [reply pytorch_toolkit/ote/ote/modules/arg_converters/base.py]: Wrote explanation

**Review by Ilya-Krylov — APPROVED** [review]

**Review by evgeny-izutov — APPROVED** [review]

**Comment by LeonidBeynenson** [comment]
2 @Ilya-Krylov : Ilya, could you merge thid PR?
CI is green.

**Pull request merged**